### PR TITLE
[RFC] Setup a way to consistently manage go versions across scripts and go.mods

### PR DESCRIPTION
**Pull request by MadhavJivrajani**
This PR does the following:

### scripts: default to using .go-version's version for tests and builds

Additionally, provide ability to opt-out of the .go-version and use a
custom one via env vars: FORCE_HOST_GO and GO_VERSION.

### .*: sync go toolchain version and add ability to verify versions

This commit adds a script to sync the version present in .go-version
across all go.mod files as the toolchain directive. As part of that,
this commit also modifies go.mod files that did not have synced toolchain
directives.

Additionally, this also adds a script to verify all toolchain and go
directives against the version present in .go-version as follows:
(1) The go directive <= version in .go-version
(2) The toolchain directive == version in .go-version

The make target for this script is run as part of `make verify` which will automatically be run as a presubmit.

/assign @ahrtr @jmhbnz @serathius

### CODE REVIEW

File: scripts/sync_go_toolchain_directive.sh
Comment (ahrtr) [inline scripts/sync_go_toolchain_directive.sh:14]: There are three options,
- only sync go line, e.g. `go 1.22.2`;
- only sync toolchain line, e.g. `toolchain go1.22.2`;
- sync both go and toolchain lines.

No matter which way we select, it will always use the highest go version by default.

For simplicity, can we go with option 1?

Comment (MadhavJivrajani) [reply scripts/sync_go_toolchain_directive.sh:14]: @ahrtr, if I understand correctly, the issue I see with option 1 and 3 is that whenever we allow bumping the `go` directive, we are also forcing developers of etcd and other projects that import etcd modules to bump their versions. From https://go.dev/doc/toolchain:
> the go line sets the minimum required Go version necessary to use a module or workspace.

File: scripts/verify_toolchain_directive.sh
Comment (ahrtr) [inline scripts/verify_toolchain_directive.sh:22]: The `go` lines in all etcd's `go.mod` files are also automatically bumped if any etcd's dependency bumped its go version. Should we also verify that the version in `go lines` <= `.go-version`?

File: scripts/sync_go_toolchain_directive.sh
Comment (ahrtr) [reply scripts/sync_go_toolchain_directive.sh:14]: Yes, it's true. It's OK to go for option 2. But please also see my next related comment https://github.com/etcd-io/etcd/pull/17876#discussion_r1582930923

**Comment by ahrtr** [comment]
Please also resolve the workflow failure: `# shellcheck disable=SC2086`

We also need to clearly document the behaviour.

File: scripts/verify_toolchain_directive.sh
Comment (MadhavJivrajani) [reply scripts/verify_toolchain_directive.sh:22]: @ahrtr that's a good point, however I'm not sure what the course of action is once the dependency bump PRs are merged. If a `go line` bump occurred and the PR is merged, do we just update the `.go-version` file then? 

My main concern with allowing dependency bumps to also bump the go line is that it can cause an issue with projects like Kubernetes that import `go.etcd.io...` modules. If the go lines of these projects are behind those of the etcd modules, then they will be either forced to bump or they cannot update the dependency till they catch up (https://github.com/kubernetes/kubernetes/issues/123744)

This behaviour is mitigated in go1.23, but till that is released and we can adopt that, it is risky allowing the go line to be bumped at all, especially from a perspective of Kubernetes consuming etcd. For bumps related to security fixes or bugs that we see in etcd, we should probably discuss what the course of action should be, but we should probably we a little more aware for usual maintenance bumps.

Comment (ahrtr) [reply scripts/verify_toolchain_directive.sh:22]: >  they cannot update the dependency till they catch up 

I think we should follow this. 

For main branch, it isn't a problem because usually we keep using the latest go version (e.g. 1.22.2 for now).

For stable releases (3.5 and 3.4 for now), usually we keep using the latest patch of previous minor version (e.g. 1.21.9 for now). Usually, we don't proactively bump dependencies for etcd stable releases unless there are any CVEs or bugs that affect etcd. If bumping a dependency for a stable release leads to the go version being automatically bumped, 
- check whether there is an alternative. For example, if both 4.20.6 and 4.21.0 of a dependency resolve a CVE, but 4.20.6 requires go1.21.9 and 4.21.0 requires go 1.22.2, then we bump 4.20.6 for etcd stable releases (3.4/3.5).
- if there is no alternative
  - if we have to bump the dependency, then we bump go version for etcd stable releases (3.4/3.5) as well.
  - if we don't have to bump the dependency (_it's unlikely running into this case_), we can hold off for now until we bump go version.


We need to clearly document the behaviour in https://github.com/etcd-io/etcd/blob/main/Documentation/contributor-guide/dependency_management.md#golang-versions

Comment (MadhavJivrajani) [reply scripts/verify_toolchain_directive.sh]: @ahrtr that sounds good, I've added the extra check.

Comment (MadhavJivrajani) [reply scripts/verify_toolchain_directive.sh]: Also, ack on documenting this.

**Comment by MadhavJivrajani** [comment]
@ahrtr on thinking about this a little more:
Whenever a dep bump changes the toolchain version, the version will be changed to whatever is the version of the local toolchain, from https://go.dev/doc/toolchain:
> For [repeatability](https://research.swtch.com/vgo-principles#repeatability), any command that updates the go line also updates the toolchain line to record its own toolchain name. 

This means that automated PRs, if it runs `./scripts/test_lib.sh`, will never accidentally change the toolchain line because we keep that in sync by setting the `GOTOOLCHAIN` env var as the version in .go-version. However, we still need to check it in case of manual bumps.

I've also changed the workflow to a presubmit instead of a periodic because now either case should not be allowed.

File: Makefile
Comment (ahrtr) [inline Makefile:201]: Please get `verify-go-versions` included in the target `verify`, and you don't add a separate workflow check because there is an existing workflow check `static-analysis.yaml`
https://github.com/etcd-io/etcd/blob/3d4d4a0053915f701e7d1b75834b309364d65109/Makefile#L55

File: Makefile
Comment (ahrtr) [inline Makefile:206]: Please also run this script in `./script/fix.sh`, so that contributors don't need to run it separately,.

File: Makefile
Comment (MadhavJivrajani) [reply Makefile:201]: Done

File: Makefile
Comment (MadhavJivrajani) [reply Makefile:206]: Done

**Comment by MadhavJivrajani** [comment]
@ahrtr can you PTAL? Thank you!

File: Makefile
Comment (ahrtr) [reply Makefile:206]: Probably I did not comment it clearly, sorry for the confusion. I wasn't saying to run `./scripts/fix.sh` in the target `sync-toolchain-directive`; instead, I was saying to run `./scripts/sync_go_toolchain_directive.sh` inside `./scripts/fix.sh`.

The `./scripts/fix.sh` is the single point of entry to fix everything for etcd.

Comment (MadhavJivrajani) [reply Makefile:206]: Ah I see, I misunderstood you there, sorry. I've made the change @ahrtr, thank you,

**Comment by ahrtr** [comment]
Overall looks good. Could you rebase this PR although github doesn't show any conflict? I see that it's still based on a commit of 2 weeks ago.

**Review by ahrtr — APPROVED** [review]
Great work!

Thanks @MadhavJivrajani 

Please also consider to update the doc as mentioned in https://github.com/etcd-io/etcd/pull/17876#discussion_r1587360706. It can be resolved in a separate PR.

**Comment by ahrtr** [comment]
cc @fuweid @ivanvc  @jmhbnz @serathius

File: scripts/fix.sh
Comment (serathius) [inline scripts/fix.sh:40]: Would prefer to move subroutines from `./scripts/fix.sh` to makefile targets. As you already have a makefile target `sync-toolchain-directive` why not add it as subtarget to `fix`?

**Review by serathius — APPROVED** [review]

File: scripts/fix.sh
Comment (ahrtr) [reply scripts/fix.sh:40]: > why not add it as subtarget to `fix`?

Yes, it's another option. Either way works for me.

We can even breakdown the existing ./script/fix.sh into a couple of Makefile targets
- fix-go-dependency (in which we will run `mod_tidy_fix` and the `sync_go_toolchain_directive.sh`)
- fix-gofmt
- fix-script-whitespace

It can be discussed & resolved in a separate PR.

**Comment by fuweid** [comment]
Hi @ahrtr @serathius @MadhavJivrajani , do we consider to apply this to library project, like bbolt and raft? since we have discussion about this in bbolt project two days ago. 

REF: https://github.com/etcd-io/bbolt/pull/747

**Comment by ahrtr** [comment]
> do we consider to apply this to library project, like bbolt and raft?

I think YES.  Please see https://github.com/etcd-io/etcd/issues/17857#issuecomment-2100965096, and please feel free to add comment in the issue.

**Comment by MadhavJivrajani** [comment]
Coming back to the discussions here this week 👍🏼

**Comment by ivanvc** [comment]
This implementation is great and simplifies what would be tedious if done by hand. We recently had the Go version updated by a new contributor, and I believe it wasn't very clear to him how to update the toolchain in the different `go.mod` files (refer to: https://github.com/etcd-io/etcd/pull/17975#issuecomment-2103331884). I think we would benefit from documenting this somewhere in the repository.

A potential improvement regarding documentation would be to document how to update the Go version. In the main branch with this new script, I think it's as easy as updating the value in `.go-version` and running `make fix` after :tada:

**Comment by jmhbnz** [comment]
> A potential improvement regarding documentation would be to document how to update the Go version. In the main branch with this new script, I think it's as easy as updating the value in `.go-version` and running `make fix` after 🎉

Should we update that into this docs page https://github.com/etcd-io/etcd/blob/main/Documentation/contributor-guide/dependency_management.md#golang-versions?

**Pull request merged**
